Modules/_operator.c

```diff
@@ -1549,16 +1549,90 @@ static PyType_Spec attrgetter_type_spec = {
 typedef struct {
     PyObject_HEAD
     PyObject *name;
-    PyObject *args;
+    PyObject *xargs; // reference to arguments passed in constructor
     PyObject *kwds;
+    PyObject **vectorcall_args;  /* Borrowed references */
+    PyObject *vectorcall_kwnames;
+    vectorcallfunc vectorcall;
 } methodcallerobject;
 
+static void * _methodcaller_initialize_vectorcall(methodcallerobject* mc)
+{
+    PyObject* args = mc->xargs;
+    PyObject* kwds = mc->kwds;
+
+    Py_ssize_t nargs = PyTuple_GET_SIZE(args);
+    mc->vectorcall_args = PyMem_Calloc(
+        nargs + (kwds ? PyDict_Size(kwds) : 0),
+        sizeof(PyObject*));
+    if (!mc->vectorcall_args) {
+        return PyErr_NoMemory();
+    }
+    /* The first item of vectorcall_args will be filled with obj later */
+    if (nargs > 1) {
+        memcpy(mc->vectorcall_args, PySequence_Fast_ITEMS(args),
+            nargs * sizeof(PyObject*));
+    }
+    if (kwds) {
+        const Py_ssize_t nkwds = PyDict_Size(kwds);
+
+        mc->vectorcall_kwnames = PyTuple_New(nkwds);
+        if (!mc->vectorcall_kwnames) {
+            return NULL;
+        }
+        Py_ssize_t i = 0, ppos = 0;
+        PyObject* key, * value;
+        while (PyDict_Next(kwds, &ppos, &key, &value)) {
+            PyTuple_SET_ITEM(mc->vectorcall_kwnames, i, Py_NewRef(key));
+            mc->vectorcall_args[nargs + i] = value; // borrowed reference
+            ++i;
+        }
+    }
+    else {
+        mc->vectorcall_kwnames = NULL;
+    }
+    return (void *)1;
+}
+
+static void _methodcaller_clear_vectorcall(methodcallerobject* mc)
+{
+    if (mc->vectorcall_args != NULL) {
+        PyMem_Free(mc->vectorcall_args);
+        mc->vectorcall_args = 0;
+        Py_CLEAR(mc->vectorcall_kwnames);
+    }
+}
```

Comment:
```suggestion
    if (mc->vectorcall_args != NULL) {
        PyMem_Free(mc->vectorcall_args);
        mc->vectorcall_args = 0;
    }
    Py_CLEAR(mc->vectorcall_kwnames);
}
```

Py_CLEAR has NULL checking.
https://github.com/python/cpython/blob/810d5d87d9fe8d86aad99e48cef4f78a72e16ccf/Include/object.h#L770-L778

It will be safer for edge cases.